master/buildbot/worker/docker.py

```diff
@@ -213,11 +213,15 @@ def _image_exists(self, client, name):
 
     def _thd_start_instance(self, image, dockerfile, volumes):
         docker_client = self._getDockerClient()
+        container_name = self.getContainerName()
         # cleanup the old instances
         instances = docker_client.containers(
             all=1,
-            filters=dict(name=self.getContainerName()))
+            filters=dict(name=container_name))
+        container_name = "/{0}".format(container_name)
         for instance in instances:
+            if container_name not in instance['Names']:
```

Comment:
The problem with codecov is that this line is not tested. Are you sure that ['Names'] is really in instance (I would have said it is ['Name']

Reply:
I did a dump and tested it locally. I was surprised that it is Names and it is a list. I did a local test to validate it works, by spinning up 20 or so builds, killing build bot and start it up again with some extra log.msg() to make sure it killed the correct containers. Not sure how we could fake this in the current test framework

Reply:
Here is a sample from the log file.

```2017-12-05 11:37:10-0600 [-] substantiating worker <LatentWorkerForBuilder builder=u'genBuilder' worker=u'docker-worker-9' state=DETACHED>
2017-12-05 11:37:10-0600 [-] ************* /buildbot-docker-worker-9-8c9f9f [u'/buildbot-docker-worker-9-8c9f9f']
2017-12-05 11:37:10-0600 [-] KILLL!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
2017-12-05 11:37:11-0600 [-] Starting factory _HTTP11ClientFactory(<function quiescentCallback at 0x7f4a500859b0>, <twisted.internet.endpoints._WrapperEndpoint object at 0x7f4a4bfe0b50>)```
oh and here is a stripped dump of the data structure we get back:
```
2017-12-05 11:24:23-0600 [-] {u'Command': u"/bin/bash --login -c 'buildbot-worker start --nodaemon'",
	 u'Created': 1512494662,
	 u'HostConfig': {u'NetworkMode': u'default'},
	 u'Id': u'36da48ba18d7c6283bdec7b32aa241379ab1a432ebaa2f56e24f2ed9fb23c4d5',
	 u'Image': u'ubuntu-17_10_bb:9ae0debffc28caf625958984f959ba1e',
	 u'ImageID': u'sha256:a8acab60e2e57240714aa3da03bad62b17e543490188ce696a686e36169f78f9',
	 u'Labels': {},
	 u'Mounts': [{u'Destination': u'...',
	              u'Mode': u'rw',
	              u'Propagation': u'rprivate',
	              u'RW': True,
	              u'Source': u'....',
	              u'Type': u'bind'}],
	 u'Names': [u'/buildbot-docker-worker-19-8c9f9f'],
	 u'NetworkSettings': {u'Networks': {u'bridge': {u'Aliases': None,
	                                                u'DriverOpts': None,
	                                                u'EndpointID': u'cf2ef9fa9dad943efa48e3818823701ea560bff4f7b74a0304c9f9fa0c209a37',
	                                                u'Gateway': u'...',
	                                                u'GlobalIPv6Address': u'...',
	                                                u'GlobalIPv6PrefixLen': 64,
	                                                u'IPAMConfig': None,
	                                                u'IPAddress': u'...',
	                                                u'IPPrefixLen': 16,
	                                                u'IPv6Gateway': u'...',
	                                                u'Links': None,
	                                                u'MacAddress': u'...',
	                                                u'NetworkID': u'ed8863131a5619a4d5de0552d46f5460edb3cfc8c82a3f46f96e8e89497e7507'}}},
	 u'Ports': [],
	 u'State': u'running',
	 u'Status': u'Up Less than a second'} 
```

Reply:
You can look at the hyper test here:
https://github.com/buildbot/buildbot/blob/master/master/buildbot/test/unit/test_worker_hyper.py#L135

I think you can easily adapt it on docker.
Here is the full commit, you will need to update the fake:
https://github.com/buildbot/buildbot/commit/29e564dfccda956b721992131b7167777aae2d0d